docs/development/design/explicit-builders.rst

```diff
@@ -0,0 +1,203 @@
+
+Some values from the context are used in our Sphinx extension,
+we should define them as configuration options instead.
+Others are used in our theme, should we stop setting them?
+
+Extension
+'''''''''
+
+There are some things our extension does that are already supported by Sphinx or themes
+(analytics, canonical URL, etc), we should write a guide instead of applying our magic.
+Alternative, we can make this options, analytics is already an option.
+
+The extension also injects some custom js and css files.
+We could try another more general approach and inject these on each HTML file after the build is complete.
+
+The extension injects the warning banner if you are reading the docs from a pull request.
+We could implement this like the version warning banner instead, so it works for MkDocs too.
+(this is injected with the version selector).
+
+MkDocs
+~~~~~~
+
+We should read the configuration file from the user or default to *one* path,
+and error if it doesn't exist.
+
+We shouldn't change the values from the user's configuration (``mkdocs.yml``),
+currently we change:
+
+- ``docs_dir``
+- ``extra_javascript``
+- ``extra_css``
+- ``google_analytics`` (we change this to ``None`` and use our own method)
+- ``theme`` (we set it to our theme for old projects)
+
+Only the additional js/css files should be added.
+Additionally, we could try another more general approach and inject these after the build is complete.
+
+Related to `#7844`_, `#4924`_, `#4827`_, `#4820`_
+
+.. _#7844: https://github.com/readthedocs/readthedocs.org/issues/7844
+.. _#4924: https://github.com/readthedocs/readthedocs.org/issues/4924
+.. _#4827: https://github.com/readthedocs/readthedocs.org/issues/4827
+.. _#4820: https://github.com/readthedocs/readthedocs.org/issues/4820
+
+Web settings
+------------
+
+Simple defaults, without fallbacks.
+
+Currently if some of our settings aren't set or are incorrect
+we try to guess some values for the user.
+We should have some sane defaults and error otherwise.
+Some are:
+
+- Requirements file (we shouldn't install any if isn't set)
+- Sphinx/MkDocs configuration file (we could default to ``docs/conf.py`` and ``mkdocs.yml``)
+
+.. note::
+
+   When using the v2 of the config file we remove all this magic.
+
+Other settings are used for things that can be done from the user side:
+
+- Analytics code (exposed as an option)
+- Canonical domain (Sphinx only, and isn't exposed as an option)
+
+Adoption
+--------
+
+If we remove some magical behaviour that was doing things for the user,
+we should document how to do them using Sphinx/MkDocs.
+
+These new builders/environments would be under a feature flag.
+We can keep the implementation incrementally by start using a feature flag on some of our projects first,
+after we everything is implemented we can move the flag to be active for projects created after ``x`` date,
+and past projects would use the old ones.
```

Comment:
I'd suggest making this new builder opt-in and use `build.builder: 2` instead of a feature flag. I'm not sure we can deprecate and remove the builder that we currently have. I think it's a good product for .com users.